### PR TITLE
Support batching multiple RPC requests

### LINKED ISSUE #186: Support batching multiple RPC requests
### Feature Request

Allow batching multiple RPC calls within the same request.
The nodes are accepting the array of RPC calls and they execute them all in parallel.

### PR DESCRIPTION
**Pull request by franciszekjob**
## Describe your changes

Suport batching multiple RPC requests
- add `StarknetRequest` and `StarknetBatchRequest` structs
- add `StarknetProvider.batchRequests()` accepting both list and variadic parameters
- add `JsonRpcParams` enum and extension
- add batch-related tests

## Linked issues

Closes #186, #145

## Breaking changes

- [x] This issue contains breaking changes

- `StarknetAccount`
   - `executeV1(...) async throws -> StarknetInvokeTransactionResponse` -> `executeV1(...) -> StarknetRequest<StarknetInvokeTransactionResponse>`
   - `executeV3(...) async throws -> StarknetInvokeTransactionResponse` -> `executeV3(...) -> StarknetRequest<StarknetInvokeTransactionResponse> `
   - `executeV1(...) async throws -> StarknetInvokeTransactionResponse` -> `executeV1(...) -> StarknetRequest<StarknetInvokeTransactionResponse>`
   - `executeV3(...) async throws -> StarknetInvokeTransactionResponse` -> `executeV3(...) -> StarknetRequest<StarknetInvokeTransactionResponse>`
   - `estimateFeeV1(...) async throws -> StarknetFeeEstimate` -> `estimateFeeV1(...) -> StarknetRequest<[StarknetFeeEstimate]>`
   - `estimateFeeV3(...) async throws -> StarknetFeeEstimate` -> `estimateFeeV3(...) -> StarknetRequest<[StarknetFeeEstimate]>`
   - `estimateDeployAccountFeeV1(...) async throws -> StarknetFeeEstimate` -> `estimateDeployAccountFeeV1(...) -> StarknetRequest<[StarknetFeeEstimate]>`
   - `estimateDeployAccountFeeV3(...) async throws -> StarknetFeeEstimate` -> `estimateDeployAccountFeeV3(...) -> StarknetRequest<[StarknetFeeEstimate]>`
   - `getNonce() async throws -> Felt` -> `getNonce()-> StarknetRequest<Felt>`
- `StarknetProvider`
   - `specVersion() async throws -> String ` -> `specVersion() -> StarknetRequest<String>`
   - `callContract(...) async throws -> [Felt]` -> `callContract(...) -> StarknetRequest<[Felt]>`
   - `estimateMessageFee(...) async throws -> StarknetFeeEstimate` -> `estimateMessageFee(...) -> StarknetRequest<StarknetFeeEstimate>`
   - `getNonce() async throws -> Felt` -> `getNonce() -> StarknetRequest<Felt>`
   - `addInvokeTransaction(...) async throws -> StarknetInvokeTransactionResponse` -> `addInvokeTransaction(...) -> StarknetRequest<StarknetInvokeTransactionResponse>` 
   - add `addDeployAccountTransaction(...) async throws -> StarknetDeployAccountResponse` ->  `addDeployAccountTransaction(...) -> StarknetRequest<StarknetDeployAccountResponse>` 
   - `getClassHashAt(...) async throws -> Felt` -> `getClassHashAt(...) -> StarknetRequest<Felt>`
   - `getBlockNumber() async throws -> UInt64` -> `getBlockNumber() -> StarknetRequest<UInt64>`
   - `getBlockHashAndNumber() async throws -> StarknetBlockHashAndNumber` -> `getBlockHashAndNumber() -> StarknetRequest<StarknetBlockHashAndNumber>`
   - `getEvents(...) async throws -> StarknetGetEventsResponse` -> `getEvents(...) -> StarknetRequest<StarknetGetEventsResponse>`
   - `getTransactionBy(...) async throws -> any StarknetTransaction` -> `getTransactionBy(...) -> StarknetRequest<StarknetTransaction>`
   - `getTransactionReceiptBy(...) async throws -> any StarknetTransactionReceipt` -> `getTransactionReceiptBy(...) -> StarknetRequest<StarknetTransactionReceipt>`
   - `getTransactionStatusBy(...) async throws -> StarknetGetTransactionStatusResponse` -> `getTransactionStatusBy(...) async throws -> StarknetRequest<StarknetGetTransactionStatusResponse>`
   - `getChainId() async throws -> StarknetChainId` -> `getChainId() -> StarknetRequest<StarknetChainId>`
   - `simulateTransactions(...) async throws -> [StarknetSimulatedTransaction]` -> `simulateTransactions(...) -> StarknetRequest<[StarknetSimulatedTransaction]>`

### CODE REVIEW

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (franciszekjob) [inline Tests/StarknetTests/Providers/ProviderTests.swift:409]: 1. This test needs to be improved - it should perform calls to get data about different transactions.
2. There should be a test for incorrect responses order. Imho it should be placed in a separate class eg. `JsonRpcResponseTests` as it focuses on the responses parsing mechanism.

Wdyt? @DelevoXDG

File: Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift
Comment (franciszekjob) [inline Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift:76]: When using `batchRequests` with variadic params, deduplication mechanism causes a possibility to receive different numbers of responses than requests. Example:
1. We send a batch request including two calls to get transaction data with the same transaction hashes.
2. We receive an array with one element.

It needs to be handled it some some. Possibly we can even delete the versions with variadic params. Wdhy? @DelevoXDG

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (THenry14) [inline Sources/Starknet/Accounts/StarknetAccount.swift:161]: should be return [0] here?

File: Sources/Starknet/Network/BatchRequest.swift
Comment (THenry14) [inline Sources/Starknet/Network/BatchRequest.swift:26]: this seems like it could be really slow for longer response lists. I suppose we need ordering to properly deserialize? is there any other way?

Comment (franciszekjob) [reply Sources/Starknet/Network/BatchRequest.swift:26]: We need to reorder the responses to match the order in which they were sent. This can be achieved with sorting `O(nlogn)` or direct placement `O(n)`. Second solution was accepted and implemented in JVM SDK, where there is same situation, [here](https://github.com/software-mansion/starknet-jvm/pull/455#discussion_r1595342132).

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (franciszekjob) [reply Sources/Starknet/Accounts/StarknetAccount.swift:161]: Yes, please look at the [old implementation](https://github.com/software-mansion/starknet.swift/pull/187/files#diff-38b6912cf986d0587267d2fcef717a429f3df0977ca75721555cada603ca4bffL216) of calling `estimateFee`. We could directly access value at index 0, because the returned type [here](https://github.com/software-mansion/starknet.swift/pull/187/files#diff-38b6912cf986d0587267d2fcef717a429f3df0977ca75721555cada603ca4bffL184) was `[StarknetFeeEstimate]`. Now it's `Request<[StarknetFeeEstimate], EstimateFeeParams>`, therefore first we need to send and then access [0] element.

File: Sources/Starknet/Network/HttpNetworkProvider.swift
Comment (franciszekjob) [reply Sources/Starknet/Network/HttpNetworkProvider.swift]: Actually I wanted to send batch request with one request inside and simply reuse `sendBatch`. However, sending single request is better in the conceptual aspect. I've refactored it - now there are two `send` methods with different signatures. 

I've also added private methods for error handling to reduce duplicated code.

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (DelevoXDG) [inline Tests/StarknetTests/Providers/ProviderTests.swift:422]: We can catch an error and verify if it's correct one

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (DelevoXDG) [reply Tests/StarknetTests/Providers/ProviderTests.swift:409]: 1. I think that can be enough
2. Yes, should add such test to `JsonRpcResponseTests`

File: Sources/Starknet/Data/Transaction/TransactionReceiptWrapper.swift
Comment (franciszekjob) [reply Sources/Starknet/Data/Transaction/TransactionReceiptWrapper.swift:3]: Yes, it's used in [`getTransactionReceiptBy()`](https://github.com/software-mansion/starknet.swift/blob/fdee240961735d03cb1ebe818d8e7419c60bc1be/Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift#L163)

File: Sources/Starknet/Network/HttpNetworkProvider.swift
Comment (franciszekjob) [reply Sources/Starknet/Network/HttpNetworkProvider.swift:50]: As [official documentation](https://developer.apple.com/documentation/swift/never) explains:

> Use Never as the return type when declaring a closure, function, or method that unconditionally throws an error, traps, or otherwise does not terminate.

Without marking `Never` here, the code would look like that:
```swift
try handleEncoding(...)
fatalError(...)
``` 

Imo using `Never` is more suitable in this case. Wdyt? @DelevoXDG

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (franciszekjob) [reply Sources/Starknet/Accounts/StarknetAccount.swift:109]: Correct, they also use provider methods. I've updated `StarknetAccount` methods to return `Request`.

File: Sources/Starknet/Network/HttpNetworkProvider.swift
Comment (franciszekjob) [reply Sources/Starknet/Network/HttpNetworkProvider.swift:74]: I also had that on my mind, but consider that `payload`  has different type in `send` methods. Because of that we would have to create two separate methods accepting `some Encodable` and `[some Encodable]`. Do you think it's worth to create them in this situation?

File: Sources/Starknet/Providers/StarknetProvider/JsonRpcPayload.swift
Comment (franciszekjob) [reply Sources/Starknet/Providers/StarknetProvider/JsonRpcPayload.swift:5]: It's unnecessary, old implementation required it to be `var`, now it's not needed.

File: Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift
Comment (franciszekjob) [reply Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift:29]: Changing access level of param structs to public requires to change access level of their encode method as well. It's a consequence of the [thread above](https://github.com/software-mansion/starknet.swift/pull/187/files#r1601988454).

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (franciszekjob) [reply Tests/StarknetTests/Providers/ProviderTests.swift]: Sure, I've added check for error message 👍

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (franciszekjob) [reply Tests/StarknetTests/Providers/ProviderTests.swift:409]: I've added [test for incorrect responses order](https://github.com/software-mansion/starknet.swift/blob/85d70fcd517c2766830124e5d6b50ee9d69f7c30/Tests/StarknetTests/Data/JsonRpcResponseTests.swift#L118)

File: Sources/Starknet/Network/BatchRequest.swift
Comment (franciszekjob) [reply Sources/Starknet/Network/BatchRequest.swift]: It was intended to be an internal constructor, as we don't want to users to create instances on their own. Due to change form class to struct in `StarknetRequest` and `StarknetBatchRequest`, this redundant initializer has been removed.

File: Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift
Comment (franciszekjob) [reply Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift:42]: I renamed it to `getHttpConfiguration` to `getHttpRequestConfiguration`, hope it's more specific. Wdyt? @DelevoXDG 
Also, I moved it to private extension as you suggested.

File: Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift
Comment (franciszekjob) [reply Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift:67]: I inlined content of `getBatchRequestRpcPayloads` there, but had to keep `getHttpConfiguration`, which is also used in `buildRequest`.

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (franciszekjob) [reply Tests/StarknetTests/Providers/ProviderTests.swift]: Sure, I restored `var provider: StarknetProviderProtocol!`, added separate `batchProvider` and used it in test for `batchRequests`.

File: Tests/StarknetTests/Accounts/AccountTest.swift
Comment (franciszekjob) [inline Tests/StarknetTests/Accounts/AccountTest.swift:118]: The [change of estimate fee return type](https://github.com/software-mansion/starknet.swift/pull/187/files#diff-27821160cc3ae3a86a78970f05cff3a582fb6894068fe88eef74e693981c9effL157-R157) resulted in having to access element `[0]` every time we use this metod. Unfortunately there isn't any other way, because first we need to send the request and the access its result.

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (franciszekjob) [inline Tests/StarknetTests/Providers/ProviderTests.swift:433]: I don't really like the fact that there is double nesting here, but there isn't any other way to access `jsonRpcError` message.

**Comment by franciszekjob** [comment]
I addressed all requested changes, also added few remark with new implementation.

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (DelevoXDG) [inline Sources/Starknet/Accounts/StarknetAccount.swift:125]: Are you sure getting the first estimate is a valid approach?

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (DelevoXDG) [inline Sources/Starknet/Accounts/StarknetAccount.swift:137]: This seems problematic to me. Someone can try to batch two `executeV1` requests like this, and they will both have the same nonce. 

I'm thinking we should think of the ways to prevent users from shooting themselves in the foot by either:
- Disallowing batching for some methods, specifically those that rely on sending other requests to generate a request
- Adding this info to docstring/README

WDYT?

File: Sources/Starknet/Network/HttpNetworkProvider.swift
Comment (DelevoXDG) [reply Sources/Starknet/Network/HttpNetworkProvider.swift:74]: Can't we just use `Any: Encodable`?

File: Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift
Comment (DelevoXDG) [reply Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift:4]: > It could be theoretically achieved, if we decrease params' level of abstraction.

Can we do that? I think it would considerably simplify the codebase (no unnecessary generis, no public param structs), but if that's too much effort it can be done in a separate PR.

File: Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift
Comment (DelevoXDG) [inline Sources/Starknet/Providers/StarknetProvider/StarknetProvider.swift:171]: Nit: 
```suggestion
    private func prepareHttpRequestConfiguration() -> HttpNetworkProvider.Configuration {
```

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (franciszekjob) [reply Sources/Starknet/Accounts/StarknetAccount.swift:125]: Has been discussed in previous thread
https://github.com/software-mansion/starknet.swift/pull/187#discussion_r1600240910

File: Sources/Starknet/Network/HttpNetworkProvider.swift
Comment (franciszekjob) [reply Sources/Starknet/Network/HttpNetworkProvider.swift:74]: Fine, we can 👍

File: Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift
Comment (franciszekjob) [reply Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift]: Yep, I've added `EncodableParams` which is a enum with possible params, there's also `EncodableParams` extension with encode method. This way we keep all param structs internal and have only one generic 👍

File: Sources/Starknet/Accounts/StarknetAccount.swift
Comment (franciszekjob) [reply Sources/Starknet/Accounts/StarknetAccount.swift]: I think disallowing it in code would be a safer choice. Maybe let's address it in another PR, as it is not blocking core functionality.

File: Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift
Comment (DelevoXDG) [inline Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift:182]: Seems to me the naming of cases is a bit inconsistent. 

Some have `Params` postfix, some `Payload`, some none. We normally don't need to include postfix that is already present in enum name, so let's drop all the postfixes:
```suggestion
    case getNonce(GetNonceParams)
    case addInvokeTransaction(AddInvokeTransactionParams)
    case wrappedExecutableTransaction(WrappedExecutableTransaction)
    case emptySequence(EmptySequence)
    case empty(EmptyParams)
    case call(CallParams)
    case estimateFee(EstimateFeeParams)
    case estimateMessageFee(EstimateMessageFeeParams)
    case addDeployAccountTransaction(AddDeployAccountTransactionParams)
    case getClassHashAt(GetClassHashAtParams)
    case getEvents(GetEventsPayload)
    case getTransactionByHash(GetTransactionByHashParams)
    case getTransactionByBlockIdAndIndex(GetTransactionByBlockIdAndIndex)
    case getTransactionReceipt(GetTransactionReceiptPayload)
    case getTransactionStatus(GetTransactionStatusPayload)
    case simulateTransactions(SimulateTransactionsParams)
```

**Review by DelevoXDG — COMMENTED** [review]
Mostly ok, few minor comments to be addressed before merging

File: Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift
Comment (franciszekjob) [reply Sources/Starknet/Providers/StarknetProvider/JsonRpcParams.swift]: 1. It has been renamed according to your suggested changes.
2. Removed.

File: Sources/Starknet/Network/StarknetBatchRequest.swift
Comment (franciszekjob) [reply Sources/Starknet/Network/StarknetBatchRequest.swift]: I renamed it to `orderRpcResults`. It wasn't inlined on purpose, because the function is used in [testBatchResponseWithIncorrectOrder](https://github.com/software-mansion/starknet.swift/blob/5cfd5366c9a8179f7e9e48ea06c69cea5e5c5698/Tests/StarknetTests/Data/JsonRpcResponseTests.swift#L146)

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (franciszekjob) [reply Tests/StarknetTests/Providers/ProviderTests.swift:34]: Yup, I added it to the protocol 👍

File: Sources/Starknet/Network/StarknetBatchRequest.swift
Comment (DelevoXDG) [inline Sources/Starknet/Network/StarknetBatchRequest.swift:17]: Nit: 

can't we use `rpcResponses.count` instead?

File: Sources/Starknet/Network/StarknetBatchRequest.swift
Comment (DelevoXDG) [inline Sources/Starknet/Network/StarknetBatchRequest.swift:26]: Nit: 

I think we should handle success first, and all the errors later.

File: Tests/StarknetTests/Providers/ProviderTests.swift
Comment (DelevoXDG) [reply Tests/StarknetTests/Providers/ProviderTests.swift:34]: If that's the case, can we just go ahead and use `provider` instead of `batchProvider`?

**Review by DelevoXDG — APPROVED** [review]
LGTM overall, can be merged after addressing the comments

**Pull request merged**
